File: R/pdist.R

guess.dim <- function(x,y)
{
   if(is.vector(x)){
      if(is.null(y)){
	     d <- 1
	  }
	  else{
	     if(is.matrix(y)){
		    d <- ncol(y)
		 }
		 else{
		    if(length(x)==length(y)){
			   d <- length(x)
			}
			else{
			   d <- 1
			}
		 }
	  }
   }
   else{
      d <- ncol(x)
   }
   d
}

pdist <- function(x,y=NULL,d,p=2,w=NULL,arc=FALSE)
{
   if(missing(d)) d <- guess.dim(x,y)
   x <- matrix(x,ncol=d)
   nx <- nrow(x)
   if(!is.null(y)){
	  y <- matrix(y,ncol=d)
	  return(pdistxy(x,y,d,p,w))
   }
   out <- rep(0,nx^2)
   if(is.infinite(p)){
	  retval <- .C("pdistinf",
				   x = as.double(t(x)),
				   n = as.integer(nx),
				   d = as.integer(d),
				   w = as.double(w),
				   out=as.double(out),PACKAGE="cccd")
   }
   else if(!is.na(p)){
	  retval <- .C("pdist",
				   x = as.double(t(x)),
				   n = as.integer(nx),
				   d = as.integer(d),
				   p = as.double(p),
				   w = as.double(w),
				   out=as.double(out),PACKAGE="cccd")
   }
	else{
	   return(pdistcos(x,d=d,arc=arc))
	}
   return(matrix(retval$out,nrow=nrow(x)))
}

pdistcos <- function(x,d=ncol(x),arc=FALSE)
{
   x <- matrix(x,ncol=d)
   nx <- nrow(x)
   out <- rep(0,nx^2)
   retval <- .C("pdistcos",
				x = as.double(t(x)),
				n = as.integer(nx),
				d = as.integer(d),
				out=as.double(out),PACKAGE="cccd")
	if(arc) retval$out <- acos(1-retval$out)
   return(matrix(retval$out,nrow=nrow(x)))
}

pdistxy <- function(x,y,d,p=2,w=NULL)
{
   if(missing(d)){
	  d <- guess.dim(x,y)
   }
   x <- matrix(x,ncol=d)
   nx <- nrow(x)
   y <- matrix(y,ncol=d)
   ny <- nrow(y)

   out <- rep(0,nx*ny)
   if(is.infinite(p)){
	  retval <- .C("pdistxyinf",
				   x = as.double(t(x)),
				   y = as.double(t(y)),
				   nx = as.integer(nx),
				   ny = as.integer(ny),
				   d = as.integer(d),
				   w = as.double(w),
				   dist=as.double(out),PACKAGE="cccd")
   }
   else {
	  retval <- .C("pdistxy",
				   x = as.double(t(x)),
				   y = as.double(t(y)),
				   nx = as.integer(nx),
				   ny = as.integer(ny),
				   d = as.integer(d),
				   p = as.double(p),
				   w = as.double(w),
				   dist=as.double(out),PACKAGE="cccd")
   }
   matrix(retval$dist,byrow=TRUE,nrow=nx)
}

File: src/pdist.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <R.h>

void pdist(double *x,int *N, int *D, double *P, double *w, double *dist)
{
   int i,j,k;
   int n=*N,d=*D;
   double p=*P;

   if(w != NULL){
	  for(i=0;i<n;i++){
		 for(j=i;j<n;j++){
			dist[i*n+j] = 0.0;
			for(k=0;k<d;k++){
			   dist[i*n+j] += pow(w[k]*fabs(x[i*d+k]-x[j*d+k]),p);
			}
			dist[i*n+j] = pow(dist[i*n+j],1./p);
			dist[j*n+i] = dist[i*n+j];
		 }
	  }
   }
   else {
	  for(i=0;i<n;i++){
		 for(j=i;j<n;j++){
			dist[i*n+j] = 0.0;
			for(k=0;k<d;k++){
			   dist[i*n+j] += pow(fabs(x[i*d+k]-x[j*d+k]),p);
			}
			dist[i*n+j] = pow(dist[i*n+j],1./p);
			dist[j*n+i] = dist[i*n+j];
		 }
	  }
   }
}

void pdistcos(double *x,int *N, int *D, double *dist)
{
   int i,j,k;
   int n=*N,d=*D;
	double l1,l2;

  for(i=0;i<n;i++){
	 for(j=i;j<n;j++){
		dist[i*n+j] = 0.0;
		l1 = l2 = 0.0;
		for(k=0;k<d;k++){
			dist[i*n+j] += x[i*d+k]*x[j*d+k];
			l1 += x[i*d+k]*x[i*d+k];
			l2 += x[j*d+k]*x[j*d+k];
		}
		if(l1>0 && l2>0){
			dist[i*n+j] /= sqrt(l1*l2);
		}
		dist[i*n+j] = 1.0-dist[i*n+j];
		dist[j*n+i] = dist[i*n+j];
	 }
  }
}

void pdistxy(double *x,double *y,int *Nx, int *Ny, int *D, double *P, 
             double *w, double *dist)
{
   int i,j,k;
   int nx=*Nx,ny=*Ny,d=*D;
   double p=*P;

   if(w != NULL){
	  for(i=0;i<nx;i++){
		 for(j=0;j<ny;j++){
			dist[i*ny+j] = 0.0;
			for(k=0;k<d;k++){
			   dist[i*ny+j] += pow(w[k]*fabs(x[i*d+k]-y[j*d+k]),p);
			}
			dist[i*ny+j] = pow(dist[i*ny+j],1./p);
		 }
	  }
   }
   else {
	  for(i=0;i<nx;i++){
		 for(j=0;j<ny;j++){
			dist[i*ny+j] = 0.0;
			for(k=0;k<d;k++){
			   dist[i*ny+j] += pow(fabs(x[i*d+k]-y[j*d+k]),p);
			}
			dist[i*ny+j] = pow(dist[i*ny+j],1./p);
		 }
	  }
   }
}

#define MAX(x,y) ((x)>(y)?(x):(y))

void pdistinf(double *x,int *N, int *D, double *w, double *dist)
{
   int i,j,k;
   int n=*N,d=*D;

   if(w != NULL){
	  for(i=0;i<n;i++){
		 for(j=i;j<n;j++){
			dist[i*n+j] = 0.0;
			for(k=0;k<d;k++){
			   dist[i*n+j] = MAX(dist[i*n+j],(w[k]*fabs(x[i*d+k]-x[j*d+k])));
			}
			dist[j*n+i] = dist[i*n+j];
		 }
	  }
   }
   else {
	  for(i=0;i<n;i++){
		 for(j=i;j<n;j++){
			dist[i*n+j] = 0.0;
			for(k=0;k<d;k++){
			   dist[i*n+j] = MAX(dist[i*n+j],fabs(x[i*d+k]-x[j*d+k]));
			}
			dist[j*n+i] = dist[i*n+j];
		 }
	  }
   }
}

void pdistxyinf(double *x,double *y,int *Nx, int *Ny, int *D, 
             double *w, double *dist)
{
   int i,j,k;
   int nx=*Nx,ny=*Ny,d=*D;

   if(w != NULL){
	  for(i=0;i<nx;i++){
		 for(j=0;j<ny;j++){
			dist[i*ny+j] = 0.0;
			for(k=0;k<d;k++){
			   dist[i*ny+j] = MAX(dist[i*ny+j],w[k]*fabs(x[i*d+k]-y[j*d+k]));
			}
		 }
	  }
   }
   else {
	  for(i=0;i<nx;i++){
		 for(j=0;j<ny;j++){
			dist[i*ny+j] = 0.0;
			for(k=0;k<d;k++){
			   dist[i*ny+j] = MAX(dist[i*ny+j],fabs(x[i*d+k]-y[j*d+k]));
			}
		 }
	  }
   }
}
